// @awesome-cordova-plugins/app-center-analytics/__ivy_ngcc__/ngx/index.js
import { __extends } from "tslib";
import { Injectable } from '@angular/core';
import { AwesomeCordovaNativePlugin, cordova } from '@awesome-cordova-plugins/core';
import * as ɵngcc0 from '@angular/core';
var AppCenterAnalytics = /** @class */ (function (_super) {
    __extends(AppCenterAnalytics, _super);
    function AppCenterAnalytics() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AppCenterAnalytics.prototype.trackEvent = function (eventName, properties) { return cordova(this, "trackEvent", {}, arguments); };
    AppCenterAnalytics.prototype.isEnabled = function () { return cordova(this, "isEnabled", {}, arguments); };
    AppCenterAnalytics.prototype.setEnabled = function (shouldEnable) { return cordova(this, "setEnabled", {}, arguments); };
    AppCenterAnalytics.pluginName = "AppCenterAnalytics";
    AppCenterAnalytics.plugin = "cordova-plugin-appcenter-analytics";
    AppCenterAnalytics.pluginRef = "AppCenter.Analytics";
    AppCenterAnalytics.repo = "https://github.com/Microsoft/appcenter-sdk-cordova/tree/master/cordova-plugin-appcenter-analytics";
    AppCenterAnalytics.platforms = ["Android", "iOS"];
AppCenterAnalytics.ɵfac = /*@__PURE__*/ function () { var ɵAppCenterAnalytics_BaseFactory; return function AppCenterAnalytics_Factory(t) { return (ɵAppCenterAnalytics_BaseFactory || (ɵAppCenterAnalytics_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(AppCenterAnalytics)))(t || AppCenterAnalytics); }; }();
AppCenterAnalytics.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: AppCenterAnalytics, factory: function (t) { return AppCenterAnalytics.ɵfac(t); } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AppCenterAnalytics, [{
        type: Injectable
    }], null, null); })();
    return AppCenterAnalytics;
}(AwesomeCordovaNativePlugin));
export { AppCenterAnalytics };
